### Summary Statistics

Summary statistics provide a concise overview of a dataset's distribution. Here's a breakdown of the commonly calculated metrics:

**Central Tendency**
- **Mean**: The average value of the dataset. It's calculated by summing all values and dividing by the number of values.
- **Median**: The middle value in a dataset when the values are arranged in ascending order. It's a robust measure that's less sensitive to outliers than the mean.
- **Mode**: The most frequent value in the dataset. There can be one mode (unimodal), multiple modes (multimodal), or no mode.

**Dispersion**
- **Standard Deviation**: Measures the spread of the data around the mean. A higher standard deviation indicates greater variability.
- **Variance**: The square of the standard deviation.
- **Range**: The difference between the maximum and minimum values.
- **Interquartile Range (IQR)**: The difference between the 75th percentile (Q3) and the 25th percentile (Q1). It measures the spread of the middle 50% of the data.
---

**Category Name: Subflow related features, Window size and TCP Flags, Flow Duration and idle time**

- Subflow Fwd Packets: The total number of forward packets within the subflow.
- Subflow Fwd Bytes: The total number of bytes transferred in the forward direction within the subflow.
- Subflow Bwd Packets: The total number of backward packets within the subflow.
- Subflow Bwd Bytes: The total number of bytes transferred in the backward direction within the subflow.
- FWD Init Win Bytes: The initial window size advertised by the forward flow.
- Bwd Init Win Bytes: The initial window size advertised by the backward flow.
- Fwd Act Data Pkts: The number of active data packets in the forward direction.
- Active Mean, Std, Max, Min: The mean, standard deviation, maximum, and minimum active times for the flow.
- Idle Mean, Std, Max, Min: The mean, standard deviation, maximum, and minimum idle times for the flow.
- Total TCP Flow Time: The total duration of the TCP flow.


In [1]:
# Importing important libraries 
import pandas as pd 
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import os 
import json 

In [2]:
# Functions 
# Update this code block to add more functions as per need
def read_data(data_path: str, usecols: list) -> pd.DataFrame:
    """
    This function reads the data from the data path
    """
    data = pd.read_csv(data_path, usecols=usecols)
    return data

**Data**

In [3]:
# Read data from filepath 
# This file path is for temporary usage (For EDA)
# It will be removed once the task is completed 

file_path = "dataset.csv"
# Columns in this category
columns_to_read = ['Subflow Fwd Packets', 'Subflow Fwd Bytes', 'Subflow Bwd Packets',
       'Subflow Bwd Bytes', 'FWD Init Win Bytes', 'Bwd Init Win Bytes',
       'Fwd Act Data Pkts', 'Fwd Seg Size Min', 'Active Mean', 'Active Std',
       'Active Max', 'Active Min', 'Idle Mean', 'Idle Std', 'Idle Max',
       'Idle Min', 'Total TCP Flow Time']

# Read the data
try:
    data = read_data(file_path, usecols = columns_to_read)
    print(f"Successfully read {len(data.columns)} features")
except:
    print("File not found!")
    print("[INFO] Please place the dataset.csv in the directory for use!")

Successfully read 17 features


**Analysis**

In [4]:
# Describe
data.describe()

Subflow Fwd Packets  Subflow Fwd Bytes  Subflow Bwd Packets  \
count         3.231475e+06       3.231475e+06         3.231475e+06   
mean          8.419746e-02       1.090405e+02         4.518061e-05   
std           2.776838e-01       2.147488e+02         6.721501e-03   
min           0.000000e+00       0.000000e+00         0.000000e+00   
25%           0.000000e+00       3.300000e+01         0.000000e+00   
50%           0.000000e+00       4.700000e+01         0.000000e+00   
75%           0.000000e+00       7.900000e+01         0.000000e+00   
max           1.000000e+00       1.054700e+04         1.000000e+00   

       Subflow Bwd Bytes  FWD Init Win Bytes  Bwd Init Win Bytes  \
count       3.231475e+06        3.231475e+06        3.231475e+06   
mean        1.064850e+02        6.023249e+04        4.957609e+02   
std         5.991156e+02        1.583170e+04        8.530959e+02   
min         0.000000e+00        0.000000e+00        0.000000e+00   
25%         1.300000e+01        6.424000e+04        5.020000e+02   
50%         3.200000e+01        6.424000e+04        5.020000e+02   
75%         3.500000e+01        6.424000e+04        5.020000e+02   
max         1.813700e+04        6.549500e+04        6.470400e+04   

       Fwd Act Data Pkts  Fwd Seg Size Min   Active Mean    Active Std  \
count       3.231475e+06      3.231475e+06  3.231475e+06  3.231475e+06   
mean        7.550082e+01      3.164003e+01  2.437296e+05  1.734073e+05   
std         4.182399e+02      5.804100e+00  2.654499e+06  1.768455e+06   
min         0.000000e+00      0.000000e+00  0.000000e+00  0.000000e+00   
25%         1.000000e+00      3.200000e+01  0.000000e+00  0.000000e+00   
50%         1.000000e+00      3.200000e+01  0.000000e+00  0.000000e+00   
75%         2.000000e+00      3.200000e+01  0.000000e+00  0.000000e+00   
max         7.830000e+03      4.000000e+01  1.148279e+08  7.718153e+07   

         Active Max    Active Min     Idle Mean      Idle Std      Idle Max  \
count  3.231475e+06  3.231475e+06  3.231475e+06  3.231475e+06  3.231475e+06   
mean   4.792189e+05  1.168187e+05  6.715270e+05  1.259603e+05  8.242769e+05   
std    4.228964e+06  2.195276e+06  5.274227e+06  1.262020e+06  5.958127e+06   
min    0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00   
25%    0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00   
50%    0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00   
75%    0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00   
max    1.148279e+08  1.148279e+08  1.199621e+08  7.711554e+07  1.199621e+08   

           Idle Min  Total TCP Flow Time  
count  3.231475e+06         3.231475e+06  
mean   5.529524e+05         7.186553e+07  
std    4.972078e+06         1.686803e+09  
min    0.000000e+00         0.000000e+00  
25%    0.000000e+00         6.991000e+03  
50%    0.000000e+00         3.671400e+04  
75%    0.000000e+00         1.232845e+05  
max    1.199621e+08         8.638926e+10

In [5]:
# unique 
data.nunique()

Subflow Fwd Packets         2
Subflow Fwd Bytes        3036
Subflow Bwd Packets         2
Subflow Bwd Bytes        1796
FWD Init Win Bytes       1976
Bwd Init Win Bytes       4781
Fwd Act Data Pkts        5371
Fwd Seg Size Min            5
Active Mean            106109
Active Std              74852
Active Max             100615
Active Min              53851
Idle Mean              116004
Idle Std                78592
Idle Max               109211
Idle Min               102036
Total TCP Flow Time    629615
dtype: int64

In [6]:
data['Subflow Bwd Packets'].value_counts()

Subflow Bwd Packets
0    3231329
1        146
Name: count, dtype: int64

> Subflow bwd packets has most values as 0 indicating there are close to none bwd packets within the subflow

In [8]:
data['Subflow Fwd Packets'].value_counts()

Subflow Fwd Packets
0    2959393
1     272082
Name: count, dtype: int64

> This shows that there are some subflow packets in Fwd subflow.

In [10]:
data['Fwd Seg Size Min'].value_counts()

Fwd Seg Size Min
32    2827826
40     274014
8       63671
0       51783
20      14181
Name: count, dtype: int64

In [13]:
data['FWD Init Win Bytes'].value_counts()

FWD Init Win Bytes
64240    2488326
65280     531312
0         127569
502        19938
501        14967
          ...   
11695          1
7746           1
7781           1
6500           1
9233           1
Name: count, Length: 1976, dtype: int64

> 64240 is the most used window size advertised in the forward flow

In [15]:
data['Active Mean'].value_counts()

Active Mean
0.0           3114433
31.0              112
27.0              110
33.0              102
25.0               96
               ...   
14862866.0          1
13479777.6          1
16929474.0          1
12551961.0          1
22329.5             1
Name: count, Length: 106109, dtype: int64

In [16]:
data['Idle Mean'].value_counts()

Idle Mean
0.000000e+00    3113423
1.100796e+07          8
1.100796e+07          8
1.075196e+07          8
1.177596e+07          8
                 ...   
6.857778e+06          1
7.115686e+06          1
6.289454e+06          1
6.151582e+06          1
2.686503e+07          1
Name: count, Length: 116004, dtype: int64

> Active and idle mean are mostly 0.